File: ch06/heap_sort.java

import java.util.ArrayList;
import java.util.Random;

class Heap {
  ArrayList<Number> v;
  int tail;

  Heap(int len) {
    if(len < 1) {
      throw new IllegalArgumentException
        ("Heap constructor: heap need length at least 1");
    }
    v = new ArrayList<>();
    tail = len;
    for(int i = 1; i <= len + 1; i = i+1) {
      v.add(0); // int auto convert to Number?
    }
  }

  //------------------------------

  private int left(int i) { return i*2; }
  private int right(int i) { return i*2+1; }
  private int parent(int i) { return i/2; }

  private boolean greater(int i, int j) {
    if (v.get(i) instanceof Integer) {
      return v.get(i).intValue() > v.get(j).intValue();
    } else {
      return v.get(i).floatValue() > v.get(j).floatValue();
    }
  }

  private boolean less(int i, int j) {
    if (v.get(i) instanceof Integer) {
      return v.get(i).intValue() < v.get(j).intValue();
    } else {
      return v.get(i).floatValue() < v.get(j).floatValue();
    }
  }

  private void exchange(int i, int j) {
    Number temp = v.get(i);
    v.set(i, v.get(j));
    v.set(j, temp);
  }

  void heapifyMax(int i) {
    int l = left(i);
    int r = right(i);
    int largest = i;

    if(l <= tail && greater(l, i)) {
      largest = l;
    }
    if(r <= tail && greater(r, largest)) {
      largest = r;
    }
    if( largest != i ) {
      exchange(i, largest);
      heapifyMax(largest);
    }
  }

  void buildMax() {
    for(int i = parent(tail); i >= 1; i = i-1) {
      heapifyMax(i);
    }
  }

  boolean isMax() {
    for(int i = 2; i <= tail; i = i+1) {
      if(less(parent(i), i)) {
        return false;
      }
    }
    return true;
  }

  void sort() {
    int tail_BK = tail;
    buildMax();
    for(tail = tail-1; tail >= 1; tail = tail-1) {
      exchange(1, tail + 1);
      heapifyMax(1);
    }
    tail = tail_BK;
  }

  Number extractPriorityQueue() {
    Number temp = v.get(1);
    exchange(1, tail);
    v.remove(tail);
    tail = tail - 1;
    heapifyMax(1);
    return temp;
  }

  void insertPriorityQueue(Number value) {
    v.add(value);
    tail = tail + 1;
    for(int i = tail;
        i <= 2 && less(parent(i), i);
        i = parent(i)) {
      exchange(i, parent(i));
    }
  }

  //------------------------------

  private Random rand = new Random();

  void setRandom(int min, int max) {
    int bound = max - min + 1;
    for(int i = 1; i <= tail; i = i+1) {
      v.set(i, (Integer)rand.nextInt(bound) + min);
    }
  }

  void setRandom() {
    for(int i = 1; i <= tail; i = i+1) {
      v.set(i, (Float)rand.nextFloat());
    }
  }

  void print() {
    System.out.print("(Heap: #" + tail + " [");
    for(int i = 1; i <= tail; i = i+1) {
      System.out.print(v.get(i) + " ");
    }
    System.out.println("])");
  }
}

class Main {
  public static void main(String[] args) {
    Heap x = new Heap(100);
    x.setRandom(0, 999);
    x.print();
    x.sort();
    x.print();
  }
}

File: ch06/heap_sort.py

import random


def setRandom(v, _b, end):
  for i in range(1, len(v)):
    v[i] = random.randint(_b, end)

def printV(v):
  print('(Heap: #', end='')
  print(len(v) - 1, end='')
  print(' [', end='')
  for i in range(1, len(v)):
    print(v[i], end='')
    print(' ', end='')
  print('])')

#-----------------------------------

def left(i):
  return i*2

def right(i):
  return i*2+1

def parent(i):
  return i//2

# v: numbers list
# i: index
# j: index
def exchange(v, i, j):
  temp = v[i]
  v[i] = v[j]
  v[j] = temp

def heapifyMax(v, _b, _e):
  l = left(_b)
  r = right(_b)
  largest = _b

  if l <= _e and v[l] > v[_b]:
    largest = l
  if r <= _e and v[r] > v[largest]:
    largest = r
  if largest != _b:
    exchange(v, largest, _b)
    heapifyMax(v, largest, _e)

def buildMax(v):
  for i in range(parent(len(v) - 1), 0, -1):
    heapifyMax(v, i, len(v) - 1)

def isMax(v):
  for i in range(len(v) - 1, 1, -1):
    if v[parent(i)] < v[i]:
      return False
  return True

def heapSort(v):
  buildMax(v)

  for _e in range(len(v) - 1, 0, -1):
    exchange(v, 1, _e)
    heapifyMax(v, 1, _e - 1)

#-----------------------------------

v = [None] * (100 + 1)
setRandom(v, 0, 999)
printV(v)
buildMax(v)
printV(v)
print(isMax(v))
heapSort(v)
printV(v)
